src/Lean/Elab/PreDefinition/PartialFixpoint/Induction.lean

```diff
@@ -66,90 +83,152 @@ private def numberNames (n : Nat) (base : String) : Array Name :=
-          let packedAdm ← PProdN.genMk (mkAdmAnd α instCCPOα) adms'
-          let hNames := numberNames infos.size "h"
-          let hTypes_hmask : Array (Expr × Array Bool) ← infos.mapIdxM fun i _info => do
-            let approxNames := infos.map fun info =>
-              match info.name with
-                | .str _ n => .mkSimple n
-                | _ => `f
-            withLocalDeclsDND (approxNames.zip types) fun approxs => do
-              let ihTypes := approxs.mapIdx fun j approx => mkApp motives[j]! approx
-              withLocalDeclsDND (ihTypes.map (⟨`ih, ·⟩)) fun ihs => do
-                let f ← PProdN.mk 0 approxs
-                let Ff := F.beta #[f]
-                let Ffi := PProdN.proj motives.size i packedType Ff
-                let t := mkApp motives[i]! Ffi
-                let t ← PProdN.reduceProjs t
-                let mask := approxs.map fun approx => t.containsFVar approx.fvarId!
-                let t ← mkForallFVars (maskArray mask approxs ++ maskArray mask ihs) t
-                pure (t, mask)
-          let (hTypes, masks) := hTypes_hmask.unzip
-          withLocalDeclsDND (hNames.zip hTypes) fun hs => do
-            let packedH ←
-              withLocalDeclD `approx packedType fun approx =>
-                let packedIHType := packedMotive.beta #[approx]
-                withLocalDeclD `ih packedIHType fun ih => do
-                  let approxs := PProdN.projs motives.size packedType approx
-                  let ihs := PProdN.projs motives.size packedIHType ih
-                  let e ← PProdN.mk 0 <| hs.mapIdx fun i h =>
-                    let mask := masks[i]!
-                    mkAppN h (maskArray mask approxs ++ maskArray mask ihs)
-                  mkLambdaFVars #[approx, ih] e
-            let e' ← mkAppOptM ``fix_induct #[α, instCCPOα, F, hmono, packedMotive, packedAdm, packedH]
-            -- Should be the type of e', but with the function definitions folded
-            let packedConclusion ← PProdN.pack 0 <| ←
-              motives.mapIdxM fun i motive => do
-                let f ← mkConstWithLevelParams infos[i]!.name
-                let fEtaExpanded ← lambdaTelescope infos[i]!.value fun ys _ =>
-                  mkLambdaFVars ys (mkAppN f ys)
-                let fInst ← eqnInfo.fixedParamPerms.perms[i]!.instantiateLambda fEtaExpanded xs
-                let fInst := fInst.eta
-                return mkApp motive fInst
-            let e' ← mkExpectedTypeHint e' packedConclusion
-            let e' ← mkLambdaFVars hs e'
-            let e' ← mkLambdaFVars adms e'
-            let e' ← mkLambdaFVars motives e'
-            let e' ← mkLambdaFVars (binderInfoForMVars := .default) (usedOnly := true) xs e'
-            let e' ← instantiateMVars e'
-            trace[Elab.definition.partialFixpoint.induction] "complete body of fixpoint induction principle:{indentExpr e'}"
-            pure e'
+    let e' ← if eqnInfo.fixpointType.all isLatticeTheoretic then
```

Comment:
Opening the fixed parameters should happen in either case, right? So maybe keep that outside the `if-then-else`.